MoreLinq.Test/TestingSequence.cs

```diff
@@ -51,59 +63,262 @@ public enum Options
-        void AssertDisposed()
+        void IDisposable.Dispose()
         {
-            if (_disposed == null)
-                return;
-            Assert.That(_disposed, Is.True, "Expected sequence to be disposed.");
-            _disposed = null;
+            if (_hasEnumerated)
+                Assert.That(_disposedCount, Is.EqualTo(_enumerationCount), ExpectedDisposal);
         }
 
         public IEnumerator<T> GetEnumerator()
         {
-            if (!IsReiterationAllowed)
-                Assert.That(_sequence, Is.Not.Null, "LINQ operators should not enumerate a sequence more than once.");
+            Assert.That(_sequence is null, Is.False, TooManyEnumerations);
 
             Debug.Assert(_sequence is not null);
 
+            Assert.That(_currentlyEnumerating, Is.False, SimultaneousEnumerations);
+            _currentlyEnumerating = true;
+
+            _hasEnumerated = true;
+
             var enumerator = _sequence.GetEnumerator().AsWatchable();
-            _disposed = false;
+            var disposed = false;
             enumerator.Disposed += delegate
             {
-                Assert.That(_disposed, Is.False, "LINQ operators should not dispose a sequence more than once.");
-                _disposed = true;
+                if (!disposed)
+                {
+                    _disposedCount++;
+                    _currentlyEnumerating = false;
+                    disposed = true;
+                }
+                else if (!_options.HasFlag(Options.AllowRepeatedDisposals))
+                {
+                    Assert.Fail(TooManyDisposals);
+                }
             };
+
             var ended = false;
             enumerator.MoveNextCalled += (_, moved) =>
             {
-                Assert.That(ended, Is.False, "LINQ operators should not continue iterating a sequence that has terminated.");
+                Assert.That(disposed, Is.False, MoveNextDisposed);
+                if (!_options.HasFlag(Options.AllowRepeatedMoveNexts))
+                    Assert.That(ended, Is.False, MoveNextCompleted);
+
                 ended = !moved;
                 MoveNextCallCount++;
             };
 
-            if (!IsReiterationAllowed)
-                _sequence = null;
+            enumerator.GetCurrentCalled += delegate
+            {
+                Assert.That(disposed, Is.False, CurrentDisposed);
+                Assert.That(ended, Is.False, CurrentCompleted);
+            };
 
+            if (++_enumerationCount == _numEnumerations)
+                _sequence = null;
             return enumerator;
         }
 
         IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+    }
+
+    [TestFixture]
+    public class TestingSequenceTest
+    {
+        [Test]
+        public void TestingSequenceShouldValidateDisposal()
+        {
+            static IEnumerable<int> InvalidUsage(IEnumerable<int> enumerable)
+            {
+                var enumerator = enumerable.GetEnumerator();
+
```

Comment:
Blank and final `yield break` is intentional: to keep this and other `InvalidUsage()` methods as similar as possible for future reading. Since this (and some other) versions require `yield break` to be valid "iterator methods", it helps to have the `yield break` on all of them, and to separate it from the core invalid behavior code.

Reply:
If it makes you feel better, 👍🏼